### ISSUE TITLE
@types/react-redux: circular reference breaks inference in InferableComponentEnhancerWithProps

### ISSUE THREAD

**Issue by Jessidhia** [DefinitelyTyped/DefinitelyTyped#26093]
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] @ryym 

It seems that the `<P extends Shared<TInjectedProps, P>>` in the `InferableComponentEnhancerWithProps` is causing `P` to always be inferred as `{}` and thus the argument to the decorator to _always_ be `Component<{}>`.

This more obviously breaks function components, where there was an attempt to fix it in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25870, but I'm not sure that was the correct approach.

---

The code below would be correctly inferred and compile before the redux@4 changes, but now the `props` parameter is inferred as only having `{ children?: ReactNode }`, and giving it an explicit type causes a parameter type error instead.

```ts
// types of State / getDesc not really important, they can be `any` fwiw
import { State } from './state'
import { getDesc } from './selectors'

function mapStateToProps (state: State, ownProps: { title: string }): { desc: string } {
  return {
    desc: getDesc(ownProps.title)
  }
}

export default connect(mapStateToProps)(
  function (props) {
    return <dl><dt>{props.title}</dt><dd>{props.desc}</dd></dl>
  }
)
```

**Comment by Kovensky** [comment]
It looks like #25870 was released, but it was an incorrect fix. I'm not yet sure what the correct fix is, but it _definitely_ involves breaking the circular reference in `<P extends Shared<TInjectedProps, P>>`.

This (anonymized) component, for example, does not get its props correctly inferred, while it did in the old redux@3 version. Only the injected props are available, but not the own props, because that's how #25870 hacked the test code into compiling:

```ts
interface Props {
  data: Data
  reason: Reason
}

export const ErrorWidget = connect(null, (dispatch: Dispatch, ownProps: Props) => ({
  onAClick () {
    return dispatch(modal.show(modal.Context.A, ownProps.data.id))
  },
  onBClick () {
    return dispatch(modal.show(modal.Context.B, ownProps.data.id))
  }
}))(
  function ErrorWidget (props) {
    const {
      title,
      urls: { mini },
    } = props.data // error: no property "data" in props

    return (
      <div className={style.container}>
        <A title={title} url={mini} onClick={props.onAClick}/>
        <B title={title} url={mini} onClick={props.onBClick}/>
      </div>
    ) // props.onAClick and props.onBClick exist just fine
  }
)
```

**Comment by Kallikrein** [comment]
Would you mind providing a repository reproducing the error ?

**Comment by Kallikrein** [comment]
the fact that your dispatch props are passed just fine means your inner component match the new option implemented in #25870, but I think TInjectedProps is missing OwnProps, hence why you can't have your data key.

The fact that you see an error with `...Shared<...>...`is probably because typescript can't match any signature and provide error for both `Component<P> | Component<TInjectedProps>`...

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25870/commits/376d4c61d91eddc63c82addebed2e8fdd7fc0235#diff-ee7ace1841a23c70c0c2b663cfae1cb7R74
```typescript
component: Component<P> | Component<TInjectedProps>
```
should be
```typescript
component: Component<P> | Component<TInjectedProps & TNeedsProps >
```

**Comment by Kallikrein** [comment]
@Kovensky

On the subject of circular reference, I see it as an issue of degree of freedom : 

You can't properly infer the inner component props and properly infer the outer component props at the same time.


The problem is that : 

* If we knew the inner component props were limited to a strict subset of stateProps & dispatchProps, we would be able to infer that the inner component props extend a Partial<> of ownProps & stateProps & dispatchProps...

* *But* the inner component props may include props that are neither in ownProps, stateProps nor dispatchProps. In this case, the outerComponent "inherits" them. The only option is to infer that the inner component props extends Shared<>, which means that none or all of the props may be in common, but for those in common the typings must match.

My PR tried to solve the second case, and #25870 tries to have the first also available.


TL;DR :
* If you know your inner component type (explicitely typed, not inferred) you can infer your outer component type.
* If you want inferred typing you can't have 'remaining props' (the ones in the inner component and not from connect()) on the outer component.

As soon as you leave simple case (à la todo-app), and want to start reusing your decorators and/or inner component and compose them in various contexts, you no longer write inline "inferred props" inner components. Maybe you could consider this option ?

If we kept the implementation that `InnerComponentProps extends OwnProps & StateProps & DispatchProps`, worded in english it would mean that the inner component props would be assignable to `OwnProps & StateProps & DispatchProps`.
So if, let's say, your component is :
```typescript
interface ErrorWidgetProps {
  data: Data;
  onAClick(): void;
  onBClick(): void;
}
const ErrorWidget: React.StatelessComponent<ErrorWidgetProps> = props => null;
```
You would get an error saying that your ErrorWidget is missing a 'reason' property, and ErrorWidgetProps don't validate `T extends ErrorWidgetProps & OwnProps`.

react-redux typings shouldn't force to leak ownProps interface 'à priori', it should be to the inner component to 'chose' wether to require this prop or not.
Sometime ownProps are relevant only for mapStateToProps logic and should not leak into the inner component interface.
It's a frequent occurence if you start working with collections and itemId's, that are 'state adress logic' and not 'state data'.

Question :

Why don't you type and implement your inner component in its own space instead of inferring as an inline argument of connect()(...)?
It's more secure and enables reusability, unit testing, storybook displays etc.

```typescript
/**
 * export this for unit testing, storybook, etc.
 */
export const ErrorWidget: React.StatelessComponent<ErrorWidgetProps> = props => (
  <div className={style.container}>
    <A title={props.title} url={props.mini} onClick={props.onAClick} />
    <B title={props.title} url={props.mini} onClick={props.onBClick} />
  </div>
);
/**
 * export for app consumption
 */
export default connect(null, mapStateToProps)(ErrorWidget);
```

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
